tests/unit/test_compress.py

```diff
@@ -0,0 +1,274 @@
+# Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import gzip
+import io
+
+import pytest
+
+from botocore.compress import compress_request
+from botocore.config import Config
+from tests import mock
+
+
+def _op_with_compression():
+    op = mock.Mock()
+    op.request_compression = {'encodings': ['gzip']}
+    op.has_streaming_input = False
+    return op
+
+
+def _op_with_multiple_compressions():
+    op = mock.Mock()
+    op.request_compression = {'encodings': ['gzip', 'gzip']}
+    op.has_streaming_input = False
+    return op
+
+
+def _op_unknown_compression():
+    op = mock.Mock()
+    op.request_compression = {'encodings': ['foo']}
+    op.has_streaming_input = None
+    return op
+
+
+def _op_without_compression():
+    op = mock.Mock()
+    op.request_compression = None
+    op.has_streaming_input = False
+    return op
+
+
+def _streaming_op_with_compression():
+    op = _op_with_compression()
+    op.has_streaming_input = True
+    streaming_shape = mock.Mock()
+    streaming_shape.metadata = {}
+    op.get_streaming_input.return_value = streaming_shape
+    return op
+
+
+def _streaming_op_with_compression_requires_length():
+    op = _streaming_op_with_compression()
+    streaming_shape = mock.Mock()
+    streaming_shape.metadata = {'requiresLength': True}
+    op.get_streaming_input.return_value = streaming_shape
+    return op
+
+
+OP_WITH_COMPRESSION = _op_with_compression()
+OP_UNKNOWN_COMPRESSION = _op_unknown_compression()
+STREAMING_OP_WITH_COMPRESSION = _streaming_op_with_compression()
+REQUEST_BODY = (
+    b'Action=PutMetricData&Version=2010-08-01&Namespace=Namespace'
+    b'&MetricData.member.1.MetricName=metric&MetricData.member.1.Unit=Bytes'
+    b'&MetricData.member.1.Value=128'
+)
+
+DEFAULT_COMPRESSION_CONFIG = Config(
+    disable_request_compression=False,
+    request_min_compression_size_bytes=10420,
+)
+COMPRESSION_CONFIG_128_BYTES = Config(
+    disable_request_compression=False,
+    request_min_compression_size_bytes=128,
+)
+COMPRESSION_CONFIG_0_BYTES = Config(
+    disable_request_compression=False,
+    request_min_compression_size_bytes=0,
+)
+
+
+def request_dict():
+    return {
+        'body': REQUEST_BODY,
+        'headers': {'foo': 'bar'},
+    }
+
+
+def request_dict_with_content_encoding_header():
+    return {
+        'body': REQUEST_BODY,
+        'headers': {'foo': 'bar', 'Content-Encoding': 'identity'},
+    }
```

Comment:
Nit: I think these can simply be constants instead of functions.

Reply:
I initially thought so too, but the changes persist between each test case. I think that would only work if these were immutable unfortunately.